src/rt/sched/behaviourcore.h

```diff
@@ -569,41 +951,174 @@ namespace verona::rt
+
+    assert(!is_wait_2pl());
 
     // This slot represents a duplicate cown, so we can ignore releasing it.
-    if (cown == nullptr)
+    if (cown() == nullptr)
+    {
+      Logging::cout() << "Duplicate cown slot " << *this << Logging::endl;
       return;
+    }
 
-    if (is_ready())
+    if (is_read_only())
     {
-      yield();
-      auto slot_addr = this;
-      // Attempt to CAS cown to null.
-      if (cown->last_slot.compare_exchange_strong(
-            slot_addr, nullptr, std::memory_order_acq_rel))
+      if (next_slot() == nullptr)
       {
-        yield();
-        Logging::cout() << "No more work for cown " << cown << Logging::endl;
-        // Success, no successor, release scheduler threads reference count.
-        shared::release(ThreadAlloc::get(), cown);
-        return;
+        auto slot_addr = this;
+        if (cown()->last_slot.compare_exchange_strong(
+              slot_addr, nullptr, std::memory_order_acq_rel))
+        {
+          Logging::cout() << *this << " Reader releasing the cown "
+                          << Logging::endl;
+          if (cown()->read_ref_count.release_read())
+          {
+            Logging::cout()
+              << *this << "Last Reader releasing the cown " << Logging::endl;
+            // Last reader
+            auto w =
+              cown()->next_writer.exchange(nullptr, std::memory_order_acq_rel);
+            if (w != nullptr)
+            {
+              Logging::cout()
+                << *this << " Last Reader waking up next writer cown "
+                << " writer " << *w << Logging::endl;
+              w->resolve();
+            }
+            yield();
+
+            // Release cown as this will be set by the new thread joining the
+            // queue.
+            Logging::cout() << *this << " Last reader No more work for cown "
+                            << Logging::endl;
+            shared::release(ThreadAlloc::get(), cown());
+          }
+          yield();
+          // Release cown as this will be set by the new thread joining the
+          // queue.
+          Logging::cout() << *this << " CAS Success No more work for cown "
+                          << Logging::endl;
+          shared::release(ThreadAlloc::get(), cown());
+          return;
+        }
+        Logging::cout() << *this
+                        << " Reader another thread extending the chain cown "
+                        << Logging::endl;
+        // If we failed, then the another thread is extending the chain
+        while (next_slot() == nullptr)
+        {
+          Systematic::yield_until(
+            [this]() { return (next_slot() != nullptr); });
+          Aal::pause();
+        }
       }
 
-      yield();
+      if (is_next_slot_writer())
+      {
+        Logging::cout() << *this << "Reader setting next writer variable "
+                        << next_behaviour() << Logging::endl;
+        cown()->next_writer = next_behaviour();
```

Comment:
Could you add a comment why this doesn't need to be exchange.  I guess that if it saw Active, then it would the write would have directly added itself.